File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.py
from math import gcd

def solution(arrayA, arrayB):
    # 각각의 최대 공약수를 구한다
    
    A = arrayA[0]
    i = 1
    while i < len(arrayA):
        A = gcd(A, arrayA[i])
        i += 1
    
    B = arrayB[0]
    i = 1
    while i < len(arrayB):
        B = gcd(B, arrayB[i])
        i += 1
        
    answer = 0
    if all(b % A != 0 for b in arrayB):
        answer = max(answer, A)
            
    if all(a % B != 0 for a in arrayA):
        answer = max(answer, B)
    
    return answer

File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.java
class Solution {
    public int solution(int[] arrayA, int[] arrayB) {
        int gcdA = arrayA[0];
        for (int a : arrayA) {
            gcdA = gcd(gcdA, a);
        }
        
        int gcdB = arrayB[0];
        for (int b : arrayB) {
            gcdB = gcd(gcdB, b);
        }
        
        
        if (gcdA == 1 && gcdB == 1) {
            return 0;
        }

        
        int answer = gcdA;
        for (int b : arrayB) {
            if (b % gcdA == 0) {
                answer = 0;
                break;
            }
        }
        
        if (answer > gcdB) {
            return answer;
        }
        
        answer = gcdB;
        for (int a : arrayA) {
            if (a % gcdB == 0) {
                answer = 0;
                break;
            }
        }
        
        return answer;
    }
    
    public int gcd(int a, int b) {
        if (b > a) {
            int temp = b;
            b = a;
            a = temp;
        }
        
        return b == 0 ? a : gcd(b, a % b);
    }
}